File: src/main/java/com/openlattice/data/DataGraphService.java
/*
 * Copyright (C) 2018. OpenLattice, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * You can contact the owner of the copyright at support@openlattice.com
 *
 */

package com.openlattice.data;

import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import com.google.common.collect.HashMultimap;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Multimaps;
import com.google.common.collect.SetMultimap;
import com.google.common.eventbus.EventBus;
import com.google.common.util.concurrent.ListenableFuture;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.IMap;
import com.openlattice.analysis.requests.TopUtilizerDetails;
import com.openlattice.data.analytics.IncrementableWeightId;
import com.openlattice.data.integration.Association;
import com.openlattice.data.integration.Entity;
import com.openlattice.data.storage.HazelcastEntityDatastore;
import com.openlattice.edm.EntitySet;
import com.openlattice.edm.type.PropertyType;
import com.openlattice.graph.core.Graph;
import com.openlattice.graph.core.objects.NeighborTripletSet;
import com.openlattice.graph.edge.EdgeKey;
import com.openlattice.hazelcast.HazelcastMap;
import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.apache.commons.collections4.keyvalue.MultiKey;
import org.apache.olingo.commons.api.edm.FullQualifiedName;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Manages the graph representation of information stored in the database.
 *
 * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
 */
public class DataGraphService implements DataGraphManager {
    private static final Logger                                   logger     = LoggerFactory
            .getLogger( DataGraphService.class );
    private final        Cache<MultiKey, IncrementableWeightId[]> queryCache = CacheBuilder.newBuilder()
            .maximumSize( 1000 )
            .expireAfterWrite( 30, TimeUnit.SECONDS )
            .build();
    private EventBus                 eventBus;
    private Graph                    lm;
    private EntityKeyIdService       idService;
    private EntityDatastore          eds;
    private IMap<UUID, EntitySet>    entitySets;
    private LoadingCache<UUID, UUID> typeIds;

    public DataGraphService(
            HazelcastInstance hazelcastInstance,
            HazelcastEntityDatastore eds,
            Graph lm,
            EntityKeyIdService ids,
            EventBus eventBus ) {
        this.lm = lm;
        this.idService = ids;
        this.eds = eds;
        this.eventBus = eventBus;

        this.entitySets = hazelcastInstance.getMap( HazelcastMap.ENTITY_SETS.name() );
        this.typeIds = CacheBuilder.newBuilder()
                .maximumSize( 100000 ) // 100K * 16 = 16000K = 16MB
                .build( new CacheLoader<UUID, UUID>() {

                    @Override
                    public UUID load( UUID key ) throws Exception {
                        return entitySets.get( key ).getEntityTypeId();
                    }
                } );
    }

    @Override
    public EntitySetData<FullQualifiedName> getEntitySetData(
            UUID entitySetId,
            LinkedHashSet<String> orderedPropertyNames,
            Map<UUID, PropertyType> authorizedPropertyTypes ) {
        return eds.getEntitySetData( entitySetId, orderedPropertyNames, authorizedPropertyTypes );
    }

    @Override
    public int deleteEntitySet( UUID entitySetId, Map<UUID, PropertyType> authorizedPropertyTypes ) {
        lm.deleteVerticesInEntitySet( entitySetId );
        return eds.deleteEntitySetData( entitySetId, authorizedPropertyTypes );
    }

    @Override
    public SetMultimap<FullQualifiedName, Object> getEntity(
            UUID entitySetId,
            UUID entityKeyId,
            Map<UUID, PropertyType> authorizedPropertyTypes ) {
        return eds
                .getEntities( entitySetId, ImmutableSet.of( entityKeyId ), authorizedPropertyTypes )
                .iterator()
                .next();
    }

    @Override public int clearEntitySet(
            UUID entitySetId, Map<UUID, PropertyType> authorizedPropertyTypes ) {
        return 0;
    }

    @Override public int clearEntities(
            UUID entitySetId, Set<UUID> entityKeyIds, Map<UUID, PropertyType> authorizedPropertyTypes ) {
        return 0;
    }

    @Override public int clearAssociations( Set<EdgeKey> key ) {
        return 0;
    }

    //TODO: Return information about delete vertices.
    @Override
    public int deleteEntities(
            UUID entitySetId,
            Set<UUID> entityKeyIds,
            Map<UUID, PropertyType> authorizedPropertyTypes ) {
        lm.deleteVertices( entitySetId, entityKeyIds );
        return eds.deleteEntities( entitySetId, entityKeyIds, authorizedPropertyTypes );
    }

    @Override
    public int deleteAssociation( Set<EdgeKey> keys, Map<UUID, PropertyType> authorizedPropertyTypes ) {
        final SetMultimap<UUID, UUID> entitySetsToEntityKeyIds = HashMultimap.create();
        for ( EdgeKey edgeKey : keys ) {
            lm.deleteEdge( edgeKey, authorizedPropertyTypes );
            entitySetsToEntityKeyIds.put( edgeKey.getEdgeEntitySetId(), edgeKey.getEdgeEntityKeyId() );
        }

        return Multimaps.asMap( entitySetsToEntityKeyIds )
                .entrySet()
                .stream()
                .mapToInt( e -> eds.deleteEntities( e.getKey(), e.getValue(), authorizedPropertyTypes ) )
                .sum();
    }

    @Override
    public Map<String, UUID> createEntities(
            UUID entitySetId,
            Map<String, SetMultimap<UUID, Object>> entities,
            Map<UUID, PropertyType> authorizedPropertyTypes ) {

        final Map<String, UUID> ids = idService.assignEntityKeyIds( entitySetId, entities.keySet() );
        final Map<UUID, SetMultimap<UUID, Object>> identifiedEntities =
                entities.entrySet()
                        .stream()
                        .collect( Collectors.toMap( e -> ids.get( e.getKey() ), Entry::getValue ) );
        eds.createEntities( entitySetId, identifiedEntities, authorizedPropertyTypes );
        return ids;
    }

    @Override public void replaceEntities(
            UUID entitySetId,
            Map<UUID, SetMultimap<UUID, Object>> entities,
            Map<UUID, PropertyType> authorizedPropertyTypes ) {

    }

    @Override public void partialReplaceEntities(
            UUID entitySetId,
            Map<UUID, SetMultimap<UUID, Object>> entities,
            Map<UUID, PropertyType> authorizedPropertyTypes ) {

    }

    @Override public void replacePropertiesInEntities(
            UUID entitySetId,
            Map<UUID, SetMultimap<UUID, Map<ByteBuffer, Object>>> replacementProperties,
            Map<UUID, PropertyType> authorizedPropertyTypes ) {

    }

    @Override
    public Map<UUID, Map<String,UUID>> integrateAssociations(
            Set<Association> associations,
            Map<UUID, Map<UUID, PropertyType>> authorizedPropertiesByEntitySet ) {

        final Map<UUID, Map<String, SetMultimap<UUID, Object>>> associationEntitiesByEntitySet = new HashMap<>();
        final Set<EntityKey> entityKeys = new HashSet<>( 2 * associations.size() );

        for ( Association association : associations ) {
            final Map<String, SetMultimap<UUID, Object>> associationEntities = associationEntitiesByEntitySet
                    .computeIfAbsent( association.getKey().getEntitySetId(), k -> new HashMap<>() );
            associationEntities.put( association.getKey().getEntityId(), association.getDetails() );
            entityKeys.add( association.getSrc() );
            entityKeys.add( association.getDst() );
        }
        final Map<EntityKey, UUID> entityKeyIds = idService.getEntityKeyIds( entityKeys );

        associationEntitiesByEntitySet
                .forEach( ( entitySetId, entities ) ->
                        createEntities( entitySetId, entities, authorizedPropertiesByEntitySet.get( entitySetId ) )
                                .forEach( ( entityId, entityKeyId ) ->
                                        entityKeyIds.put( new EntityKey( entitySetId, entityId ), entityKeyId ) ) );

        associations
                .parallelStream()
                .map( association -> {
                    UUID edgeId = entityKeyIds.get( association.getKey() );

                    UUID srcId = idService.getEntityKeyId( association.getSrc() );
                    UUID srcTypeId = typeIds.getUnchecked( association.getSrc().getEntitySetId() );
                    UUID srcSetId = association.getSrc().getEntitySetId();
                    UUID dstId = idService.getEntityKeyId( association.getDst() );
                    UUID dstTypeId = typeIds.getUnchecked( association.getDst().getEntitySetId() );
                    UUID dstSetId = association.getDst().getEntitySetId();
                    UUID edgeTypeId = typeIds.getUnchecked( association.getKey().getEntitySetId() );
                    UUID edgeSetId = association.getKey().getEntitySetId();

                    return lm
                            .addEdgeAsync( srcId,
                                    srcTypeId,
                                    srcSetId,
                                    dstId,
                                    dstTypeId,
                                    dstSetId,
                                    edgeId,
                                    edgeTypeId,
                                    edgeSetId );
                } ).forEach( DataGraphService::tryGetAndLogErrors );
        return null;
    }

    @Override
    public IntegrationResults integrateEntitiesAndAssociations(
            Set<Entity> entities,
            Set<Association> associations,
            Map<UUID, Map<UUID, PropertyType>> authorizedPropertiesByEntitySetId ) {
        // Map<EntityKey, UUID> idsRegistered = new HashMap<>();

        final Map<UUID, Map<String, SetMultimap<UUID, Object>>> entitiesByEntitySet = new HashMap<>();

        for ( Entity entity : entities ) {
            final Map<String, SetMultimap<UUID, Object>> entitiesToCreate = entitiesByEntitySet
                    .computeIfAbsent( entity.getEntitySetId(), k -> new HashMap<>() );
            entitiesToCreate.put( entity.getEntityId(), entity.getDetails() );
        }

        entitiesByEntitySet
                .forEach( ( entitySetId, entitySet ) ->
                        createEntities( entitySetId,
                                entitySet,
                                authorizedPropertiesByEntitySetId.get( entitySetId ) ) );

        integrateAssociations( associations,authorizedPropertiesByEntitySetId );
        return null;
    }

    @Override
    public Stream<SetMultimap<FullQualifiedName, Object>> getTopUtilizers(
            UUID entitySetId,
            List<TopUtilizerDetails> topUtilizerDetailsList,
            int numResults,
            Map<UUID, PropertyType> authorizedPropertyTypes ) {
        /*
         * ByteBuffer queryId; try { queryId = ByteBuffer.wrap( ObjectMappers.getSmileMapper().writeValueAsBytes(
         * topUtilizerDetailsList ) ); } catch ( JsonProcessingException e1 ) { logger.debug(
         * "Unable to generate query id." ); return null; }
         */
        IncrementableWeightId[] maybeUtilizers = queryCache
                .getIfPresent( new MultiKey( entitySetId, topUtilizerDetailsList ) );
        final IncrementableWeightId[] utilizers;
        // if ( !eds.queryAlreadyExecuted( queryId ) ) {
        if ( maybeUtilizers == null ) {
            //            utilizers = new TopUtilizers( numResults );
            SetMultimap<UUID, UUID> srcFilters = HashMultimap.create();
            SetMultimap<UUID, UUID> dstFilters = HashMultimap.create();

            topUtilizerDetailsList.forEach( details -> {
                ( details.getUtilizerIsSrc() ? srcFilters : dstFilters ).
                        putAll( details.getAssociationTypeId(), details.getNeighborTypeIds() );

            } );
            utilizers = lm.computeGraphAggregation( numResults, entitySetId, srcFilters, dstFilters );

            queryCache.put( new MultiKey( entitySetId, topUtilizerDetailsList ), utilizers );
        } else {
            utilizers = maybeUtilizers;
        }
        //TODO: this returns unsorted data.
        final UUID[] utilizerIds = new UUID[ utilizers.length ];
        for ( int i = 0; i < utilizers.length; ++i ) {
            utilizerIds[ i ] = utilizers[ i ].getId();
        }
        return eds.getEntities( entitySetId, ImmutableSet.copyOf( utilizerIds ), authorizedPropertyTypes );
    }

    @Override
    public NeighborTripletSet getNeighborEntitySets( UUID entitySetId ) {
        return lm.getNeighborEntitySets( entitySetId );
    }

    public static void tryGetAndLogErrors( ListenableFuture<?> f ) {
        try {
            f.get();
        } catch ( InterruptedException | ExecutionException e ) {
            logger.error( "Future execution failed.", e );
        }
    }
}

File: src/main/kotlin/com/openlattice/data/DataGraphService.kt
/*
 * Copyright (C) 2018. OpenLattice, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * You can contact the owner of the copyright at support@openlattice.com
 *
 *
 */

package com.openlattice.data

import com.google.common.collect.ListMultimap
import com.google.common.collect.Multimaps
import com.google.common.collect.SetMultimap
import com.google.common.eventbus.EventBus
import com.google.common.util.concurrent.ListenableFuture
import com.openlattice.analysis.AuthorizedFilteredNeighborsRanking
import com.openlattice.analysis.requests.FilteredNeighborsRankingAggregation
import com.openlattice.data.integration.Association
import com.openlattice.data.integration.Entity
import com.openlattice.edm.type.PropertyType
import com.openlattice.graph.core.GraphService
import com.openlattice.graph.core.NeighborSets
import com.openlattice.graph.edge.Edge
import com.openlattice.postgres.streams.PostgresIterable
import org.apache.commons.lang3.tuple.Pair
import org.apache.olingo.commons.api.edm.FullQualifiedName
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer
import java.util.*
import java.util.concurrent.ExecutionException
import java.util.stream.Stream
import kotlin.collections.HashMap
import kotlin.collections.HashSet

/**
 *
 * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
 */
private val logger = LoggerFactory.getLogger(DataGraphService::class.java)

open class DataGraphService(
        private val eventBus: EventBus,
        private val graphService: GraphService,
        private val idService: EntityKeyIdService,
        private val eds: EntityDatastore

) : DataGraphManager {
    override fun getEntityKeyIds(entityKeys: Set<EntityKey>): Set<UUID> {
        return idService.reserveEntityKeyIds(entityKeys)
    }


    //TODO: Move to a utility class
    companion object {

        @JvmStatic
        fun tryGetAndLogErrors(f: ListenableFuture<*>) {
            try {
                f.get()
            } catch (e: InterruptedException) {
                logger.error("Future execution failed.", e)
            } catch (e: ExecutionException) {
                logger.error("Future execution failed.", e)
            }
        }

        const val ASSOCIATION_SIZE = 30000
    }


    /* Select */

    override fun getEntitySetData(
            entityKeyIds: Map<UUID, Optional<Set<UUID>>>,
            orderedPropertyNames: LinkedHashSet<String>,
            authorizedPropertyTypes: Map<UUID, Map<UUID, PropertyType>>,
            linking: Boolean
    ): EntitySetData<FullQualifiedName> {
        return eds.getEntities(entityKeyIds, orderedPropertyNames, authorizedPropertyTypes, linking)
    }

    override fun getLinkingEntitySetSize(linkedEntitySetIds: Set<UUID>): Long {
        if (linkedEntitySetIds.isEmpty()) {
            return 0
        }

        return eds.getLinkingEntities(
                linkedEntitySetIds.map { it to Optional.empty<Set<UUID>>() }.toMap(),
                mapOf()
        ).count()
    }

    override fun getEntitySetSize(entitySetId: UUID): Long {
        return eds.getEntitySetSize(entitySetId)
    }

    override fun getEntity(
            entitySetId: UUID,
            entityKeyId: UUID,
            authorizedPropertyTypes: Map<UUID, PropertyType>
    ): SetMultimap<FullQualifiedName, Any> {
        return eds.getEntities(
                entitySetId,
                setOf(entityKeyId),
                mapOf(entitySetId to authorizedPropertyTypes)
        ).iterator().next()
    }

    override fun getLinkingEntity(
            entitySetIds: Set<UUID>,
            entityKeyId: UUID,
            authorizedPropertyTypes: Map<UUID, Map<UUID, PropertyType>>
    ): SetMultimap<FullQualifiedName, Any> {
        return eds.getLinkingEntities(
                entitySetIds.map { it to Optional.of(setOf(entityKeyId)) }.toMap(),
                authorizedPropertyTypes
        ).iterator().next()
    }

    override fun getNeighborEntitySets(entitySetIds: Set<UUID>): List<NeighborSets> {
        return graphService.getNeighborEntitySets(entitySetIds)
    }

    override fun getNeighborEntitySetIds(entitySetIds: Set<UUID>): Set<UUID> {
        return getNeighborEntitySets(entitySetIds)
                .flatMap { listOf(it.srcEntitySetId, it.edgeEntitySetId, it.dstEntitySetId) }
                .toSet()
    }

    override fun getEdgesAndNeighborsForVertex(entitySetId: UUID, entityKeyId: UUID): Stream<Edge> {
        return graphService.getEdgesAndNeighborsForVertex(entitySetId, entityKeyId)
    }

    override fun getEdgeKeysOfEntitySet(entitySetId: UUID): PostgresIterable<DataEdgeKey> {
        return graphService.getEdgeKeysOfEntitySet(entitySetId)
    }

    override fun getEdgesConnectedToEntities(entitySetId: UUID, entityKeyIds: Set<UUID>)
            : PostgresIterable<DataEdgeKey> {
        return graphService.getEdgeKeysContainingEntities(entitySetId, entityKeyIds)
    }


    /* Delete */

    private val groupEdges: (List<DataEdgeKey>) -> Map<UUID, Set<UUID>> = { edges ->
        edges.map { it.edge }.groupBy { it.entitySetId }.mapValues { it.value.map { it.entityKeyId }.toSet() }
    }

    override fun clearEntitySet(entitySetId: UUID, authorizedPropertyTypes: Map<UUID, PropertyType>): WriteEvent {
        // clear edges
        val verticesCount = graphService.clearVerticesInEntitySet(entitySetId)

        //clear entities
        val entityWriteEvent = eds.clearEntitySet(entitySetId, authorizedPropertyTypes)

        logger.info("Cleared {} entities and {} vertices.", entityWriteEvent.numUpdates, verticesCount)
        return entityWriteEvent
    }

    override fun clearEntities(
            entitySetId: UUID,
            entityKeyIds: Set<UUID>,
            authorizedPropertyTypes: Map<UUID, PropertyType>
    ): WriteEvent {
        return clearEntityDataAndVertices(entitySetId, entityKeyIds, authorizedPropertyTypes)
    }

    override fun clearAssociationsBatch(
            entitySetId: UUID,
            associationsEdgeKeys: PostgresIterable<DataEdgeKey>,
            authorizedPropertyTypes: Map<UUID, Map<UUID, PropertyType>>
    ): List<WriteEvent> {
        var associationDeleteCount = 0
        val writeEvents = ArrayList<WriteEvent>()

        associationsEdgeKeys.asSequence().chunked(ASSOCIATION_SIZE, groupEdges).forEach { entityKeyIds ->
            entityKeyIds.entries.forEach {
                val writeEvent = clearEntityDataAndVertices(it.key, it.value, authorizedPropertyTypes.getValue(it.key))
                writeEvents.add(writeEvent)
                associationDeleteCount += writeEvent.numUpdates
            }
        }

        logger.info("Cleared {} associations when deleting entities from entity set {}", associationDeleteCount,
                entitySetId)

        return writeEvents
    }

    private fun clearEntityDataAndVertices(entitySetId: UUID,
                                           entityKeyIds: Set<UUID>,
                                           authorizedPropertyTypes: Map<UUID, PropertyType>): WriteEvent {
        // clear edges
        val verticesCount = graphService.clearVertices(entitySetId, entityKeyIds)

        //clear entities
        val entityWriteEvent = eds.clearEntities(entitySetId, entityKeyIds, authorizedPropertyTypes)

        logger.info("Cleared {} entities and {} vertices.", entityWriteEvent.numUpdates, verticesCount)
        return entityWriteEvent
    }

    override fun clearEntityProperties(
            entitySetId: UUID, entityKeyIds: Set<UUID>, authorizedPropertyTypes: Map<UUID, PropertyType>
    ): WriteEvent {
        val propertyWriteEvent = eds.clearEntityData(entitySetId, entityKeyIds, authorizedPropertyTypes)
        logger.info("Cleared properties {} of {} entities.",
                authorizedPropertyTypes.values.map(PropertyType::getType), propertyWriteEvent.numUpdates)

        return propertyWriteEvent
    }

    override fun deleteEntitySet(entitySetId: UUID, authorizedPropertyTypes: Map<UUID, PropertyType>): WriteEvent {
        // delete edges
        val verticesCount = graphService.deleteVerticesInEntitySet(entitySetId)

        // delete entities
        val entityWriteEvent = eds.deleteEntitySetData(entitySetId, authorizedPropertyTypes)

        logger.info("Deleted {} entities and {} vertices.", entityWriteEvent.numUpdates, verticesCount)
        return entityWriteEvent
    }

    override fun deleteEntities(
            entitySetId: UUID,
            entityKeyIds: Set<UUID>,
            authorizedPropertyTypes: Map<UUID, PropertyType>
    ): WriteEvent {
        return deleteEntityDataAnVertices(entitySetId, entityKeyIds, authorizedPropertyTypes)
    }

    private fun deleteEntityDataAnVertices(
            entitySetId: UUID,
            entityKeyIds: Set<UUID>,
            authorizedPropertyTypes: Map<UUID, PropertyType>): WriteEvent {
        // delete edges
        val entityWriteEvent = eds.deleteEntities(entitySetId, entityKeyIds, authorizedPropertyTypes)

        // delete entities
        val verticesCount = graphService.deleteVertices(entitySetId, entityKeyIds)

        logger.info("Deleted {} entities and {} vertices.", entityWriteEvent.numUpdates, verticesCount)

        return entityWriteEvent
    }

    override fun deleteAssociationsBatch(
            entitySetId: UUID,
            associationsEdgeKeys: PostgresIterable<DataEdgeKey>,
            authorizedPropertyTypes: Map<UUID, Map<UUID, PropertyType>>): List<WriteEvent> {
        var associationDeleteCount = 0
        val writeEvents = ArrayList<WriteEvent>()

        associationsEdgeKeys.asSequence().chunked(ASSOCIATION_SIZE, groupEdges).forEach { entityKeyIds ->
            entityKeyIds.entries.forEach {
                val writeEvent = deleteEntityDataAnVertices(it.key, it.value, authorizedPropertyTypes.getValue(it.key))
                writeEvents.add(writeEvent)
                associationDeleteCount += writeEvent.numUpdates
            }
        }

        logger.info("Deleted {} associations when deleting entities from entity set {}", associationDeleteCount,
                entitySetId)

        return writeEvents
    }

    override fun deleteEntityProperties(
            entitySetId: UUID,
            entityKeyIds: Set<UUID>,
            authorizedPropertyTypes: Map<UUID, PropertyType>
    ): WriteEvent {
        val propertyCount = eds.deleteEntityProperties(entitySetId, entityKeyIds, authorizedPropertyTypes)

        logger.info("Deleted properties {} of {} entities.",
                authorizedPropertyTypes.values.map(PropertyType::getType), propertyCount.numUpdates)
        return propertyCount
    }


    /* Create */

    override fun integrateEntities(
            entitySetId: UUID,
            entities: Map<String, Map<UUID, Set<Any>>>,
            authorizedPropertyTypes: Map<UUID, PropertyType>
    ): Map<String, UUID> {
        //We need to fix this to avoid remapping. Skipping for expediency.
        return doIntegrateEntities(entitySetId, entities, authorizedPropertyTypes)
                .map { it.key.entityId to it.value }.toMap()
    }

    private fun doIntegrateEntities(
            entitySetId: UUID,
            entities: Map<String, Map<UUID, Set<Any>>>,
            authorizedPropertyTypes: Map<UUID, PropertyType>
    ): Map<EntityKey, UUID> {
        val ids = idService.getEntityKeyIds(entities.keys.map { EntityKey(entitySetId, it) }.toSet())
        val identifiedEntities = ids.map { it.value to entities[it.key.entityId] }.toMap()
        eds.integrateEntities(entitySetId, identifiedEntities, authorizedPropertyTypes)

        return ids
    }

    override fun createEntities(
            entitySetId: UUID,
            entities: List<Map<UUID, Set<Any>>>,
            authorizedPropertyTypes: Map<UUID, PropertyType>
    ): Pair<List<UUID>, WriteEvent> {
        val ids = idService.reserveIds(entitySetId, entities.size)
        val entityMap = ids.mapIndexed { i, id -> id to entities[i] }.toMap()
        val writeEvent = eds.createOrUpdateEntities(entitySetId, entityMap, authorizedPropertyTypes)

        return Pair.of(ids, writeEvent)
    }

    override fun mergeEntities(
            entitySetId: UUID,
            entities: Map<UUID, Map<UUID, Set<Any>>>,
            authorizedPropertyTypes: Map<UUID, PropertyType>
    ): WriteEvent {
        return eds.createOrUpdateEntities(entitySetId, entities, authorizedPropertyTypes)
    }

    override fun replaceEntities(
            entitySetId: UUID,
            entities: Map<UUID, Map<UUID, Set<Any>>>,
            authorizedPropertyTypes: Map<UUID, PropertyType>
    ): WriteEvent {
        return eds.replaceEntities(entitySetId, entities, authorizedPropertyTypes)
    }

    override fun partialReplaceEntities(
            entitySetId: UUID,
            entities: Map<UUID, Map<UUID, Set<Any>>>,
            authorizedPropertyTypes: Map<UUID, PropertyType>
    ): WriteEvent {
        return eds.partialReplaceEntities(entitySetId, entities, authorizedPropertyTypes)
    }

    override fun replacePropertiesInEntities(
            entitySetId: UUID,
            replacementProperties: Map<UUID, SetMultimap<UUID, Map<ByteBuffer, Any>>>,
            authorizedPropertyTypes: Map<UUID, PropertyType>
    ): WriteEvent {
        return eds.replacePropertiesInEntities(entitySetId, replacementProperties, authorizedPropertyTypes)
    }

    override fun createAssociations(associations: Set<DataEdgeKey>): WriteEvent {
        return graphService.createEdges(associations)
    }

    override fun createAssociations(
            associations: ListMultimap<UUID, DataEdge>,
            authorizedPropertiesByEntitySetId: Map<UUID, Map<UUID, PropertyType>>
    ): Map<UUID, CreateAssociationEvent> {

        val associationCreateEvents: MutableMap<UUID, CreateAssociationEvent> = mutableMapOf()

        Multimaps
                .asMap(associations)
                .forEach {
                    val entitySetId = it.key
                    val entities = it.value.map { it.data }
                    val (ids, entityWrite) = createEntities(entitySetId, entities, authorizedPropertiesByEntitySetId[entitySetId]!!)

                    val edgeKeys = it.value.asSequence().mapIndexed { index, dataEdge ->
                        DataEdgeKey(dataEdge.src, dataEdge.dst, EntityDataKey(entitySetId, ids[index]))
                    }.toSet()
                    val edgeWrite = graphService.createEdges(edgeKeys)

                    associationCreateEvents[entitySetId] = CreateAssociationEvent(ids, entityWrite, edgeWrite)
                }

        return associationCreateEvents
    }

    override fun integrateAssociations(
            associations: Set<Association>,
            authorizedPropertiesByEntitySet: Map<UUID, Map<UUID, PropertyType>>
    ): Map<UUID, Map<String, UUID>> {
        val associationsByEntitySet = associations.groupBy { it.key.entitySetId }
        val entityKeys = HashSet<EntityKey>(3 * associations.size)
        val entityKeyIds = HashMap<EntityKey, UUID>(3 * associations.size)

        //Create the entities for the association and build list of required entity keys
        val integrationResults = associationsByEntitySet
                .asSequence()
                .map {
                    val entitySetId = it.key
                    val entities = it.value.asSequence()
                            .map {
                                entityKeys.add(it.src)
                                entityKeys.add(it.dst)
                                it.key.entityId to it.details
                            }.toMap()
                    val ids = doIntegrateEntities(entitySetId, entities, authorizedPropertiesByEntitySet.getValue(entitySetId))
                    entityKeyIds.putAll(ids)
                    entitySetId to ids.asSequence().map { it.key.entityId to it.value }.toMap()
                }.toMap()
        //Retrieve the src/dst keys
        idService.getEntityKeyIds(entityKeys, entityKeyIds)

        //Create graph structure.
        val edges = associations
                .asSequence()
                .map { association ->
                    val srcId = entityKeyIds[association.src]
                    val dstId = entityKeyIds[association.dst]
                    val edgeId = entityKeyIds[association.key]

                    val src = EntityDataKey(association.src.entitySetId, srcId)
                    val dst = EntityDataKey(association.dst.entitySetId, dstId)
                    val edge = EntityDataKey(association.key.entitySetId, edgeId)

                    DataEdgeKey(src, dst, edge)
                }
                .toSet()
        graphService.createEdges(edges)

        return integrationResults
    }

    override fun integrateEntitiesAndAssociations(
            entities: Set<Entity>,
            associations: Set<Association>,
            authorizedPropertiesByEntitySetId: Map<UUID, Map<UUID, PropertyType>>
    ): IntegrationResults? {
        val entitiesByEntitySet = HashMap<UUID, MutableMap<String, MutableMap<UUID, MutableSet<Any>>>>()

        for (entity in entities) {
            val entitiesToCreate = entitiesByEntitySet.getOrPut(entity.entitySetId) { mutableMapOf() }
            val entityDetails = entitiesToCreate.getOrPut(entity.entityId) { entity.details }
            if (entityDetails !== entity.details) {
                entity.details.forEach { propertyTypeId, values ->
                    entityDetails.getOrPut(propertyTypeId) { mutableSetOf() }.addAll(values)
                }
            }
        }

        entitiesByEntitySet
                .forEach { entitySetId, entitySet ->
                    integrateEntities(
                            entitySetId,
                            entitySet,
                            authorizedPropertiesByEntitySetId[entitySetId]!!
                    )
                }

        integrateAssociations(associations, authorizedPropertiesByEntitySetId)
        return null
    }

    /* Top utilizers */

    override fun getFilteredRankings(
            entitySetIds: Set<UUID>,
            numResults: Int,
            filteredRankings: List<AuthorizedFilteredNeighborsRanking>,
            authorizedPropertyTypes: Map<UUID, Map<UUID, PropertyType>>,
            linked: Boolean,
            linkingEntitySetId: Optional<UUID>
    ): Iterable<Map<String, Any>> {
//        val maybeUtilizers = queryCache
//                .getIfPresent(MultiKey(entitySetIds, filteredRankings))
//        val utilizers: PostgresIterable<Map<String, Object>>
//
//
//        if (maybeUtilizers == null) {

        return graphService.computeTopEntities(
                numResults,
                entitySetIds,
                authorizedPropertyTypes,
                filteredRankings,
                linked,
                linkingEntitySetId
        )

//            queryCache.put(MultiKey(entitySetIds, filteredRankings), utilizers)
//        } else {
//            utilizers = maybeUtilizers
//        }

//        val entities = eds
//                .getEntities(entitySetIds.first(), utilizers.map { it.id }.toSet(), authorizedPropertyTypes)
//                .map { it[ID_FQN].first() as UUID to it }
//                .toList()
//                .toMap()

//        return utilizers.map {
//            val entity = entities[it.id]!!
//            entity.put(COUNT_FQN, it.weight)
//            entity
//        }.stream()


    }

    override fun getTopUtilizers(
            entitySetId: UUID,
            filteredNeighborsRankingList: List<FilteredNeighborsRankingAggregation>,
            numResults: Int,
            authorizedPropertyTypes: Map<UUID, PropertyType>
    ): Stream<SetMultimap<FullQualifiedName, Any>> {
//        val maybeUtilizers = queryCache
//                .getIfPresent(MultiKey(entitySetId, filteredNeighborsRankingList))
//        val utilizers: Array<IncrementableWeightId>
//
//
//        if (maybeUtilizers == null) {
//            utilizers = graphService.computeTopEntities(
//                    numResults, entitySetId, authorizedPropertyTypes, filteredNeighborsRankingList
//            )
//            //            utilizers = new TopUtilizers( numResults );
//            val srcFilters = HashMultimap.create<UUID, UUID>()
//            val dstFilters = HashMultimap.create<UUID, UUID>()
//
//            val associationPropertyTypeFilters = HashMultimap.create<UUID, Optional<SetMultimap<UUID, RangeFilter<Comparable<Any>>>>>()
//            val srcPropertyTypeFilters = HashMultimap.create<UUID, Optional<SetMultimap<UUID, RangeFilter<Comparable<Any>>>>>()
//            val dstPropertyTypeFilters = HashMultimap.create<UUID, Optional<SetMultimap<UUID, RangeFilter<Comparable<Any>>>>>()
//            filteredNeighborsRankingList.forEach { details ->
//                val associationSets = edm.getEntitySetsOfType(details.associationTypeId).map { it.id }
//                val neighborSets = edm.getEntitySetsOfType(details.neighborTypeId).map { it.id }
//
//                associationSets.forEach {
//                    (if (details.utilizerIsSrc) srcFilters else dstFilters).putAll(it, neighborSets)
//                    (if (details.utilizerIsSrc) srcPropertyTypeFilters else dstPropertyTypeFilters).putAll(
//                            it, details.neighborFilters
//                    )
//                }
//
//            }
//
//            utilizers = graphService.computeGraphAggregation(numResults, entitySetId, srcFilters, dstFilters)
//
//            queryCache.put(MultiKey(entitySetId, filteredNeighborsRankingList), utilizers)
//        } else {
//            utilizers = maybeUtilizers
//        }
//
//        val entities = eds
//                .getEntities(entitySetId, utilizers.map { it.id }.toSet(), authorizedPropertyTypes)
//                .map { it[ID_FQN].first() as UUID to it }
//                .toList()
//                .toMap()
//
//        return utilizers.map {
//            val entity = entities[it.id]!!
//            entity.put(COUNT_FQN, it.weight)
//            entity
//        }.stream()
        return Stream.empty()
    }
}